### PR TITLE
Add colorbars to plots of gallery and tutorial examples

### LINKED ISSUE #2030: Missing colorbars in examples
In two examples a colormap is used but the corresponding colorbar is missing.

- [Creating a map with contour lines](https://www.pygmt.org/dev/tutorials/advanced/contour_map.html)
- [Sampling along tracks](https://www.pygmt.org/dev/gallery/images/track_sampling.html#sphx-glr-gallery-images-track-sampling-py)

This limits the interpretability of the  figures (especially for [Sampling along tracks](https://www.pygmt.org/dev/gallery/images/track_sampling.html#sphx-glr-gallery-images-track-sampling-py)).  Thus, I think a colorbar should be added.

### PR DESCRIPTION
**Pull request by yvonnefroehlich**
**Description of proposed changes**
This PR aims to add colorbars to some plots of the gallery and tutorial examples:
- [x] [Creating a map with contour lines](https://www.pygmt.org/dev/tutorials/advanced/contour_map.html) (contour_map.py)
- [ ] ~[Sampling along tracks](https://www.pygmt.org/dev/gallery/images/track_sampling.html) (track_sampling.py)~
- [x] [Plotting a surface](https://www.pygmt.org/dev/gallery/3d_plots/grdview_surface.html) (grdview_surface.py)

Fixes #2030 

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by yvonnefroehlich** [comment]
I extended issue #2030 and added also to [Plotting a surface](https://www.pygmt.org/dev/gallery/3d_plots/grdview_surface.html) a colorbar. Furthermore, I added a grid. Of course, we can discuss these changes!

--------------
Strictly speaking, two colorbars have to be added to [Sampling along tracks](https://www.pygmt.org/dev/gallery/images/track_sampling.html), as there are color-codings for both the bathymetry ("gray") and the sampled bathymetry points ("terra"). Should I add both colorbars or only the one for the sampled bathymetry points?

**Comment by seisman** [comment]
> I extended issue #2030 and added also to [Plotting a surface](https://www.pygmt.org/dev/gallery/3d_plots/grdview_surface.html) a colorbar. Furthermore, I added a grid. Of course, we can discuss these changes!

Adding the grid looks good to me.

> Strictly speaking, two colorbars have to be added to [Sampling along tracks](https://www.pygmt.org/dev/gallery/images/track_sampling.html), as there are color-codings for both the bathymetry ("gray") and the sampled bathymetry points ("terra"). Should I add both colorbars or only the one for the sampled bathymetry points?

Yes, two colorbars makes more sense. It's also a good example to show how to add two colorbars on the same map.

**Comment by seisman** [comment]
@yvonnefroehlich Thanks for your active contributions to the PyGMT project. On behalf of the PyGMT maintainers @GenericMappingTools/pygmt-maintainers, I’d like to invite you to join the “PyGMT Devs” Slack channel, which is used usually for some internal discussions. Please leave your email address here or send your email address to me (seisman.info#gmail.com) if you’d like to join.

**Comment by yvonnefroehlich** [comment]
> @yvonnefroehlich Thanks for your active contributions to the PyGMT project. On behalf of the PyGMT maintainers @GenericMappingTools/pygmt-maintainers, I’d like to invite you to join the “PyGMT Devs” Slack channel, which is used usually for some internal discussions. Please leave your email address here or send your email address to me (seisman.info#gmail.com) if you’d like to join.

Thanks @GenericMappingTools/gmt-maintainers for this positive feedback 🙂!

**Comment by yvonnefroehlich** [comment]
Hm. Currently, I am wondering why the `Style Checks` are failing just after merging `main` into this feature branch. Before doing so, they have passed. Furthermore, it seems that there are similar problems with the `Style Checks` in my other PRs (#2226, #2251, and #2252).
I noticed that running `make format` locally on `main` also gives me failing `Style Checks`:
```
(pygmt) yfroe@gpiseis16:~/Documents/01_GitHub/github_pygmt/pygmt> make format
isort .
Skipped 94 files
docformatter --in-place pygmt doc/conf.py examples
black pygmt doc/conf.py examples
All done! ✨ 🍰 ✨
249 files left unchanged.
blackdoc pygmt doc/conf.py examples
reformatted pygmt/helpers/decorators.py
reformatted pygmt/helpers/testing.py
All done! ✨ 🍰 ✨
2 files reformatted, 247 files left unchanged.
```

File: examples/tutorials/advanced/contour_map.py
Comment (willschlitzer) [inline examples/tutorials/advanced/contour_map.py:68]: I'm not sure why "the" should be removed here.

File: examples/tutorials/advanced/contour_map.py
Comment (willschlitzer) [inline examples/tutorials/advanced/contour_map.py:91]:
```suggestion
# :meth:`pygmt.Figure.grdcontour` method. Finally, a colorbar is added using the
# :meth:`pygmt.Figure.colorbar` method.
```

File: examples/tutorials/advanced/contour_map.py
Comment (yvonnefroehlich) [reply examples/tutorials/advanced/contour_map.py:68]: For me, this sounds better 🙂. But if the others agree that the current version is better/correct, I am fine with changing this back.

File: examples/tutorials/advanced/contour_map.py
Comment (yvonnefroehlich) [reply examples/tutorials/advanced/contour_map.py]: Good spotting - Thanks @willschlitzer!

File: examples/tutorials/advanced/contour_map.py
Comment (seisman) [inline examples/tutorials/advanced/contour_map.py:108]:
```suggestion
fig.colorbar(frame=["x+lelevation", "y+lm"])
```

File: examples/tutorials/advanced/contour_map.py
Comment (willschlitzer) [reply examples/tutorials/advanced/contour_map.py:68]: Since "projection" and "frame" are both singular, I would think it would make more sense to add "the".

File: examples/gallery/images/track_sampling.py
Comment (yvonnefroehlich) [inline examples/gallery/images/track_sampling.py:66]: Here two different colormaps are set up and saved in files. @GenericMappingTools/pygmt-maintainers Maybe it is better to deleted these files at the end of the example, e.g., please see this tutorial https://www.pygmt.org/dev/tutorials/basics/text.html#plotting-from-a-text-file?

File: examples/tutorials/advanced/contour_map.py
Comment (michaelgrund) [reply examples/tutorials/advanced/contour_map.py:68]: Agree Witz @willschlitzer to leave the "the" as is.

Comment (yvonnefroehlich) [reply examples/tutorials/advanced/contour_map.py]: Thanks for your answers! Changed it back in commit  https://github.com/GenericMappingTools/pygmt/pull/2190/commits/9c76c2d062abd345790c617d19147e8d63e74d31.

File: examples/gallery/images/track_sampling.py
Comment (yvonnefroehlich) [reply examples/gallery/images/track_sampling.py:69]: I have made corresponding changes in commit https://github.com/GenericMappingTools/pygmt/pull/2190/commits/4e29ce5101157a3ffc61a05d6db23ac161685789.

File: examples/gallery/images/track_sampling.py
Comment (seisman) [inline examples/gallery/images/track_sampling.py:40]:
```suggestion
    series=[-1, 1, 0.01],  # for normalized values
```

**Comment by yvonnefroehlich** [comment]
@seisman I just saw you changed to milestone of this PR from v.0.8.0 to v.0.9.0 some days ago. Which changes do you feel are needed for this  PR 🙂?

**Comment by seisman** [comment]
> @seisman I just saw you changed to milestone of this PR from v.0.8.0 to v.0.9.0 some days ago. Which changes do you feel are needed for this PR 🙂?

I can't remember why I made the change 😕. Perhaps because I don't like the changes in `track_sampling.py` (creating two CPT files and then delete). 

I thought the following code should work, but it doesn't work as I expected:
```
fig = pygmt.Figure()
fig.basemap(region="g", projection="Cyl_stere/150/-20/15c", frame=True)
fig.grdimage(grid=grid, cmap="gray")
fig.coast(land="#666666")
fig.colorbar(
    position="JBC+o0c/1.2c+ml",  # Place colorbar at position Bottom Center
    frame=["af", "x+lelevation", "y+lm"],
)

fig.plot(
    x=track.longitude,
    y=track.latitude,
    style="c0.15c",
    cmap="terra",
    fill=(track.bathymetry - track.bathymetry.mean()) / track.bathymetry.std(),
)
fig.colorbar(
    position="JRM+ml",  # Place colorbar at position Right Middle
    frame=["a0.2f0.1", "+lnormalized elevation"],
)

fig.show()
```

**Comment by yvonnefroehlich** [comment]
> I can't remember why I made the change 😕. Perhaps because I don't like the changes in `track_sampling.py` (creating two CPT files and then deleting them).

No worries.
Yep, it's not really elegant. Maybe the suggestion below is a bit better?

```python
import pygmt

# Load sample grid and point datasets
grid = pygmt.datasets.load_earth_relief()
points = pygmt.datasets.load_sample_data(name="ocean_ridge_points")
# Sample the bathymetry along the world's ocean ridges at specified track
# points
track = pygmt.grdtrack(points=points, grid=grid, newcolname="bathymetry")

fig = pygmt.Figure()
# Plot the Earth relief grid on Cylindrical Stereographic projection, masking
# land areas
fig.basemap(region="g", projection="Cyl_stere/150/-20/15c", frame=True)

# Plot Earth relief grid and add corresponding colorbar
pygmt.makecpt(cmap="gray", series=[int(grid.min()), int(grid.max()), 10])
fig.grdimage(grid=grid, cmap=True)
fig.colorbar(
    cmap=True,
    position="JBC+o0c/1.2c+ml",  # Place colorbar at position Bottom Center
    frame=["af", "x+lelevation", "y+lm"],
)

fig.coast(land="#666666")

# Plot the sampled bathymetry points using circles (c) with a diameter of
# 0.15 centimeters (c). Points are colored using elevation values (normalized
# for visual purposes)
pygmt.makecpt(cmap="terra", series=[-1, 1, 0.01])  # Set up for normalized values
fig.plot(
    x=track.longitude,
    y=track.latitude,
    style="c0.15c",
    cmap=True,
    fill=(track.bathymetry - track.bathymetry.mean()) / track.bathymetry.std(),
)
fig.colorbar(
    cmap=True,
    position="JRM+ml",  # Place colorbar at position Right Middle
    frame=["a0.2f0.1", "+lnormalized elevation"],
)

fig.show()
```

**Comment by seisman** [comment]
> > I can't remember why I made the change 😕. Perhaps because I don't like the changes in `track_sampling.py` (creating two CPT files and then deleting them).
> 
> No worries. Yep, it's not really elegant. Maybe the suggestion below is a bit better?
> 
> ```python
> import pygmt
> 
> # Load sample grid and point datasets
> grid = pygmt.datasets.load_earth_relief()
> points = pygmt.datasets.load_sample_data(name="ocean_ridge_points")
> # Sample the bathymetry along the world's ocean ridges at specified track
> # points
> track = pygmt.grdtrack(points=points, grid=grid, newcolname="bathymetry")
> 
> fig = pygmt.Figure()
> # Plot the Earth relief grid on Cylindrical Stereographic projection, masking
> # land areas
> fig.basemap(region="g", projection="Cyl_stere/150/-20/15c", frame=True)
> 
> # Plot Earth relief grid and add corresponding colorbar
> pygmt.makecpt(cmap="gray", series=[int(grid.min()), int(grid.max()), 10])
> fig.grdimage(grid=grid, cmap=True)
> fig.colorbar(
>     cmap=True,
>     position="JBC+o0c/1.2c+ml",  # Place colorbar at position Bottom Center
>     frame=["af", "x+lelevation", "y+lm"],
> )
> 
> fig.coast(land="#666666")
> 
> # Plot the sampled bathymetry points using circles (c) with a diameter of
> # 0.15 centimeters (c). Points are colored using elevation values (normalized
> # for visual purposes)
> pygmt.makecpt(cmap="terra", series=[-1, 1, 0.01])  # Set up for normalized values
> fig.plot(
>     x=track.longitude,
>     y=track.latitude,
>     style="c0.15c",
>     cmap=True,
>     fill=(track.bathymetry - track.bathymetry.mean()) / track.bathymetry.std(),
> )
> fig.colorbar(
>     cmap=True,
>     position="JRM+ml",  # Place colorbar at position Right Middle
>     frame=["a0.2f0.1", "+lnormalized elevation"],
> )
> 
> fig.show()
> ```

This version looks much better.

**Comment by yvonnefroehlich** [comment]
> > > I can't remember why I made the change 😕. Perhaps because I don't like the changes in `track_sampling.py` (creating two CPT files and then deleting them).
> > 
> > 
> > No worries. Yep, it's not really elegant. Maybe the suggestion below is a bit better?

> This version looks much better.

I changed the example `track_sampling.py` corresponding to this version.

File: examples/gallery/images/track_sampling.py
Comment (seisman) [inline examples/gallery/images/track_sampling.py:37]: `cmap=True` is not necessary for `Figure.colorbar()`:
```suggestion
```

File: examples/gallery/images/track_sampling.py
Comment (seisman) [inline examples/gallery/images/track_sampling.py:60]:
```suggestion
```

File: examples/gallery/images/track_sampling.py
Comment (seisman) [inline examples/gallery/images/track_sampling.py:62]:
```suggestion
    frame=["a0.2f0.1", "x+lnormalized elevation"],
```

**Review by seisman — APPROVED** [review]

**Review by michaelgrund — APPROVED** [review]

**Comment by yvonnefroehlich** [comment]
Hm. Maybe it be fair to ask @weiji14 whether he is fine with the changes of `track_sampling.py` as he is the original author of this example? But not sure if he is currently available 🙂.

**Review by willschlitzer — APPROVED** [review]

File: examples/gallery/images/track_sampling.py
Comment (weiji14) [inline examples/gallery/images/track_sampling.py:32]: Would [`grd2cpt`](https://www.pygmt.org/v0.7.0/api/generated/pygmt.grd2cpt.html) be better here instead of `makecpt`? We actually don't have an example for `grd2cpt` yet :slightly_smiling_face:

File: examples/gallery/images/track_sampling.py
Comment (weiji14) [inline examples/gallery/images/track_sampling.py:62]: Usually I do like colorbars, but in this case, the normalized elevations are hard to interpret, so I'm not sure if adding a colorbar for it is helpful. If it were raw elevation values in metres, the colorbar would make sense, but then the colors of the points won't look nice :sweat_smile:

File: examples/gallery/images/track_sampling.py
Comment (weiji14) [inline examples/gallery/images/track_sampling.py:43]: Adding the shorelines makes the map look a bit more complicated, maybe it's better to remove them?

**Review by weiji14 — COMMENTED** [review]
> Hm. Maybe it would be fair to ask @weiji14 whether he is fine with the changes of `track_sampling.py` as he is the original author of this example? But not sure if he is currently available slightly_smiling_face. _Edit_: I think three approving reviewers should be enough for merging.

Sorry for the late response. The original intention of the `track_sampling.py` gallery example added in https://github.com/GenericMappingTools/pygmt/pull/308 was to showcase the `grdtrack` data processing function, rather than as a proper 'map' per se.

Personally, I feel that having just one colorbar (the normalized elevation one) would be enough, since the background grayscale earth_relief elevation is just for context. But I also find normalized elevation to be hard to interpret (see my comment below), so not sure what's a good plan forward :sweat_smile: What are your thoughts?

File: examples/gallery/images/track_sampling.py
Comment (yvonnefroehlich) [reply examples/gallery/images/track_sampling.py:43]: I added the shorelines, to emphasise that only the elevation of the water masses is color-coded by the gray colormap, but the land masses are colored with a fixed gray hue.

File: examples/gallery/images/track_sampling.py
Comment (yvonnefroehlich) [reply examples/gallery/images/track_sampling.py:32]: Thanks for this suggestions! Can have a look at it.

File: examples/gallery/images/track_sampling.py
Comment (yvonnefroehlich) [reply examples/gallery/images/track_sampling.py:62]: > Usually I do like colorbars, but in this case, the normalized elevations are hard to interpret, so I'm not sure if adding a colorbar for it is helpful.

Yes, the numerical values are hard to interpret. But without a colorbar, one does not know to which quantity the color-coding referes and that the values were normalized.

> If it were raw elevation values in metres, the colorbar would make sense, but then the colors of the points won't look nice 😅

Yep, I tried this 😆.

File: examples/gallery/images/track_sampling.py
Comment (yvonnefroehlich) [reply examples/gallery/images/track_sampling.py:32]: I feel I can not finish the work on changing this example regarding this comment today to get this PR into the v0.8.0 release.

**Comment by seisman** [comment]
> The original intention of the `track_sampling.py` gallery example added in #308 was to showcase the `grdtrack` data processing function, rather than as a proper 'map' per se.
> 
> Personally, I feel that having just one colorbar (the normalized elevation one) would be enough, since the background grayscale earth_relief elevation is just for context. But I also find normalized elevation to be hard to interpret (see my comment below), so not sure what's a good plan forward 😅 What are your thoughts?

Perhaps we can rewrite the grdtrack example to something like (https://docs.gmt-china.org/latest/examples/ex026/):

**Comment by weiji14** [comment]
> Perhaps we can rewrite the grdtrack example to something like (https://docs.gmt-china.org/latest/examples/ex026/)

Yes, how about we revert the changes to track_sampling and redesign the example in a separate PR? I think the other changes that add a colorbar are good to go.

**Comment by yvonnefroehlich** [comment]
> > Perhaps we can rewrite the grdtrack example to something like (https://docs.gmt-china.org/latest/examples/ex026/)
> 
> Yes, how about we revert the changes to track_sampling and redesign the example in a separate PR? I think the other changes that add a colorbar are good to go.

That's fine with me.

**Comment by seisman** [comment]
> > Perhaps we can rewrite the grdtrack example to something like ([docs.gmt-china.org/latest/examples/ex026](https://docs.gmt-china.org/latest/examples/ex026/))
> 
> Yes, how about we revert the changes to track_sampling and redesign the example in a separate PR? I think the other changes that add a colorbar are good to go.

Sounds good.

**Comment by yvonnefroehlich** [comment]
I reverted the changes of `track_sampling.py` via commit https://github.com/GenericMappingTools/pygmt/pull/2190/commits/ad2bd7957911f5eb8ac3e6e907b7e145a899c7f7.

**Review by seisman — APPROVED** [review]

**Review by weiji14 — APPROVED** [review]
Thanks @yvonnefroehlich, and sorry for the back and forth :slightly_smiling_face:

**Pull request merged**
